sympy/functions/elementary/tests/test_piecewise.py

```diff
@@ -534,8 +534,13 @@ def test_piecewise_simplify():
         ).simplify() == Piecewise((1, (x >= 2) & (x < oo)), (nan, True))
     assert Piecewise((1, x < 2), (2, (x > 1) & (x < 3)), (3, True)
         ). simplify() == Piecewise((1, x < 2), (2, x < 3), (3, True))
-    assert Piecewise((1, x < 2), (2, (x <= 3) & (x > 1)), (3, True)
-        ).simplify() == Piecewise((1, x < 2), (2, x <= 3), (3, True))
+
+    # See test_piecewise_simplify_xfail_redundant_inequality below
+    #
+    # assert Piecewise((1, x < 2), (2, (x <= 3) & (x > 1)), (3, True)
+    #     ).simplify() == Piecewise((1, x < 2), (2, x <= 3), (3, True))
+    #
```

Comment:
I don't think that we should leave commented-out test because people in the future cannot grasp the idea about why it was commented out.
I'd simply suggest to remove this, until passing the review.

Reply:
> people in the future cannot grasp the idea about why it was commented out.

The comment points to the function below which explains this:
```python
@XFAIL
def test_piecewise_simplify_xfail_redundant_inequality():
    # Uncomment this case in test_piecewise_simplify() when it is fixed.
    # See https://github.com/sympy/sympy/pull/25605
    assert Piecewise((1, x < 2), (2, (x <= 3) & (x > 1)), (3, True)
        ).simplify() == Piecewise((1, x < 2), (2, x <= 3), (3, True))
```

Reply:
But why does the test need to be commented in instead of removing xfail?

I think that the code can anyway change in the future, and the commented out code may be unsearchable

Reply:
> But why does the test need to be commented in instead of removing xfail?

I don't quite understand what this sentence means.

In future ideally the commented out code would be uncommented again because that test will pass again but some change somewhere else would be needed first. We could just delete the code now but the commented out code shows what the code should be expected to look like in future which makes it easier for someone to fix it again when possible. If someone in future discovers that the XFAIL test passes then there are comments left that can helpfully explain exactly how to restore the test code to how it should be.

> I think that the code can anyway change in the future, and the commented out code may be unsearchable

The comments in both places reference each other by naming the respective functions. They are searchable because they have the exact same tested code.

Reply:
I meant that someone can just remove the xfail check in after changing the code to be successful

Reply:
I also don’t really get that why the test example cannot be fixed, or the test expression can be changed to reflect the weaker simplify.

It just seems like we trade another failing test case with fix, which may not be an good way to trade.

Reply:
> The relationship between the change here and the failed test is not obvious

Actually this is not true: the relationship is obvious. The buggy code that is changed here is responsible for that simplification. In the case of the failed test here the simplification is correct but in the case of the bug report in gh-25603 the same code gives an incorrect result. The fact that the test passes at all on master is only because there is buggy code that might sometimes return a correct result by luck.

What is not obvious is if there ever was any code in SymPy that could perform this simplification correctly without also causing wrong results half of the time. Further investigation would be needed to determine if a simple change to the code could make it better at simplification without giving wrong results but it is not trivial because the code in question looks at one interval at a time but the correctness depends on the other intervals.

Reply:
I don’t have the strong opinion about my suggestion change. But I don’t agree that your idea is best in every respect though. There are tradeoffs, and I rarely had adopted that direction in my work.

> When someone does that I would want them to put the test code back where it was rather than leaving different cases scattered in a disorganised way.

It is still technically cleanly achievable by git reverting specific git commit though.
However, people don’t put much emphasis on git commits anyway.

Reply:
And although testing against exact snapshots of simplify can give lots of false negatives, it is a good idea if we want to approach test cases more conservatively.

Reply:
> I don’t have the strong opinion about pushing my suggestion.

It is not clear to me from the comments above what your suggestion (singular?) actually is. You seem to have suggested many different things that have no discernible pattern like:

- Just delete the test
- Just comment out the test and not have any XFAIL test
- Delete the current test code and make an XFAIL test instead
- Change the test to reflect the new output (even though the output is not fully simplified and the test is a test of simplification).
- Just fix it so that the test passes (i.e. invent a new simplification algorithm that previously did not exist).

The only thing I can conclude from seeing all these suggestions and comments is that you want something different somehow but it is not clear to me what it is that you dislike or why you would consider any or all of these alternatives to be better.

It is possible that I would agree with you if I understood what point  you were trying to make but either way as far as I can tell we are discussing something of very minor importance in the context of this particular PR. If the point is more general than this PR then I think that it would be best made elsewhere i.e. please open a separate issue to discuss how to handle cases like this in general and please clearly explain and motivate what your suggestion is. Then if we can agree a clear "policy" we will not need to rehash this discussion in future.

This PR is a very simple PR that fixes a very obvious bug. I have marked it as easy to fix so that a relatively new contributor can make a PR to fix it. We should not hold up something like this over minor things. If there is a change needed to the PR then it needs to be explained very clearly and carefully what that change should be and why it should be made.

In the context of SymPy we operate on the principle of "consensus" which basically means that a PR should not be merged if anyone objects to it. That means that everyone needs to be clear all the time about whether everyone else does or does not object to things. If you want to make minor tangential points that sort of look like objections then you should also be very clear about whether you do or do not object to the PR being merged (otherwise the comments are frustrating for other contributors). Likewise if you do object then it is important to be clear precisely what it is that you object to and what are the alternatives that you would not object to.

Reply:
I don’t object

Reply:
> I don’t object

Thank you.

If you want to open a separate issue then we can discuss in more general terms what is a good policy for handling cases like this. It sounds like you have some ideas from other contexts that might be good for SymPy but I think it would be better to discuss them elsewhere.

Reply:
> it is that you dislike or why you would consider any or all of these alternatives to be better.

I consider all alternatives to be compromising something and not better really.
I’d like to wait a certain period time until someone gives better fix though,
We have lots of backlogs of bugs, but I am now determined that we don’t and can’t have to fix everything swiftly.

Reply:
There should be a way in which the succeeding conditions on domains of x, take into account the negation of all preceding conditions. So the prev master function had the preceding condition of `(x>2)` thus the next condition of `((x<=3)&(x>1)) `could might as well be written as `((x<=3) & (x>1) & (!x>2))`

In the PR example, since the boolean is in the first condition, it doesn't really make a difference. Does this seem logically sound?

In case the piecewise was as follows : `Piecewise((1, x>4), (log(x), (x <= 5) & (x > 3)), (x, True))` that would mean
1, when x>4
log(x), when (x<=5) & (x>3) & (!x>4)   =>   (x>3) & (x<=4) 
x, in all other cases i.e. (x<=3)

Reply:
> Does this seem logically sound?

I consider that logically sound though. Simplify can be lazy.

If it is easy to follow up what’s wrong about the logic in brain, 
I’d expect implementing that shouldn’t be that difficult by code. 
But if there is some logic easy to reason in brain, but difficult to write in code, that is another problem that I’m interested to figure out though.

Reply:
> In the PR example, since the boolean is in the first condition, it doesn't really make a difference. Does this seem logically sound?

Yes, I think so. I guess a bigger change to the code that does this would be needed so that simplify could handle this properly without bugs.